### PR TITLE
feat(iast): include as iast telemetry metrics tags in the root span of the trace

### PR DESCRIPTION
**Pull request by avara1986**
For metrics that need to be correlated to traces, a small subset of IAST telemetry metrics can be included as tags in the root span of the trace following the next algorithm, this PR contains:
- `iast_metrics` into asm_request_context
- Span metrics with IAST `request.tainted` and `executed.skins.*` data

TODO: add integration tests with a flask server and retrieve telemetry events from test agent but we need to refactor telemetry tests and move testagent fixture from conftest to a upper level conftest

## Checklist

- [x] Change(s) are motivated and described in the PR description.
- [x] Testing strategy is described if automated tests are not included in the PR.
- [x] Risk is outlined (performance impact, potential for breakage, maintainability, etc).
- [x] Change is maintainable (easy to change, telemetry, documentation).
- [x] [Library release note guidelines](https://ddtrace.readthedocs.io/en/stable/releasenotes.html) are followed. If no release note is required, add label `changelog/no-changelog`.
- [x] Documentation is included (in-code, generated user docs, [public corp docs](https://github.com/DataDog/documentation/)).
- [x] Backport labels are set (if [applicable](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting))

## Reviewer Checklist

- [x] Title is accurate.
- [x] No unnecessary changes are introduced.
- [x] Description motivates each change.
- [x] Avoids breaking [API](https://ddtrace.readthedocs.io/en/stable/versioning.html#interfaces) changes unless absolutely necessary.
- [x] Testing strategy adequately addresses listed risk(s).
- [x] Change is maintainable (easy to change, telemetry, documentation).
- [x] Release note makes sense to a user of the library.
- [x] Reviewer has explicitly acknowledged and discussed the performance implications of this PR as reported in the benchmarks PR comment.
- [x] Backport labels are set in a manner that is consistent with the [release branch maintenance policy](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting)
- [x] If this PR touches code that signs or publishes builds or packages, or handles credentials of any kind, I've requested a review from `@DataDog/security-design-and-guidance`.
- [x] This PR doesn't touch any of that.

### CODE REVIEW

File: ddtrace/appsec/_iast/_metrics.py [hunk 1]
```diff
@@ -98,10 +102,49 @@ def _set_metric_iast_executed_sink(vulnerability_type):
...
+@metric_verbosity(TELEMETRY_INFORMATION_VERBOSITY)
+def _set_metric_iast_request_tainted():
+    total_objects_tainted = _request_tainted()
     if total_objects_tainted > 0:
         telemetry_writer.add_count_metric(TELEMETRY_NAMESPACE_TAG_IAST, "request.tainted", total_objects_tainted)
+
+
+def _set_span_tag_iast_request_tainted(span):
+    total_objects_tainted = _request_tainted()
+    print("_set_span_tag_iast_request_tainted!!!!!!!!!!!!!!!!!!!!1")
```
Comment (gnufede) [inline ddtrace/appsec/_iast/_metrics.py:120]:
```suggestion
```

File: ddtrace/appsec/_iast/_metrics.py [hunk 2]
```diff
@@ -98,10 +102,49 @@ def _set_metric_iast_executed_sink(vulnerability_type):
...
+    print(total_objects_tainted)
```
Comment (gnufede) [inline ddtrace/appsec/_iast/_metrics.py:121]:
```suggestion
```

File: tests/appsec/iast/test_telemetry.py [hunk 3]
```diff
@@ -98,10 +107,24 @@ def test_metric_request_tainted(mock_telemetry_lifecycle_writer):
...
+    with flask_server(iast_enabled="true", token=None) as context:
+        _, flask_client, pid = context
+
+        response = flask_client.get("/iast-cmdi-vulnerability?filename=path_traversal_test_file.txt")
+
+        assert response.status_code == 200
+        assert response.content == b"OK"
+
+    # TODO: move tests/telemetry/conftest.py::test_agent_session into a common conftest
+    #   events = test_agent_session.get_events()
```
Comment (gnufede) [inline tests/appsec/iast/test_telemetry.py:130]: Left behind comment?

**Review by juanjux — DISMISSED** [review]

File: ddtrace/appsec/_iast/_metrics.py [hunk 4]
```diff
@@ -98,10 +102,48 @@ def _set_metric_iast_executed_sink(vulnerability_type):
...
+    data[full_key] = result + counter
+
+
+def get_iast_span_metrics() -> Dict:
+    return _IAST_SPAN_METRICS
+
+
+def reset_iast_span_metrics() -> None:
+    global _IAST_SPAN_METRICS
+    _IAST_SPAN_METRICS = dict()
```
Comment (christophe-papazian) [inline ddtrace/appsec/_iast/_metrics.py:149]:
```suggestion
    _IAST_SPAN_METRICS.clear()
```
It doesn't create an additional object/reference and release the memory immediatly

**Review by christophe-papazian — APPROVED** [review]

**Review by christophe-papazian — APPROVED** [review]

**Review by juanjux — APPROVED** [review]

**Review by gnufede — APPROVED** [review]

**Review by emmettbutler — APPROVED** [review]

**Review by christophe-papazian — APPROVED** [review]

**Review by juanjux — APPROVED** [review]

**Pull request merged**